test/integration/cpo_tso_assign.py

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python3
+
+'''
+MIT License
+
+Copyright (c) 2021 Futurewei Cloud
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+'''
+
+import argparse, unittest, sys, os, signal, time
+import requests, json
+from urllib.parse import urlparse
+import shlex, subprocess
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--tso_child_pid", help="TSO child PID")
+parser.add_argument("--prometheus_port", help="CPO prometheus port")
+parser.add_argument("--cmd", help="Command to run the new TSOs")
+args = parser.parse_args()
+
+def readMetrics():
+    url = "http://127.0.0.1:" + args.prometheus_port + "/metrics"
+    r = requests.get(url)
+    success_count = 0
+    failure_count = 0
+    for line in r.text.splitlines():
+        if "CPOService_CPO_succeeded_tso_assignments" in line:
+            try:
+                success_count = int(float(line.split()[1]))
+            except:
+                continue
+            print("CPOService_CPO_succeeded_tso_assignments: ", success_count)
+        
+        if "CPOService_CPO_failed_tso_assignments" in line:
+            print("CPOService_CPO_failed_tso_assignments: ", line)
+            try:
+                failure_count = int(float(line.split()[1]))
+                print("CPOService_CPO_failed_tso_assignments: ", failure_count)
+            except:
+                continue
+    return success_count, failure_count
+
+
+
+class TestTSOAssignFailure(unittest.TestCase):
+    ### testing TSO assignment failure. Start with unreachable TSOs. Confirm the retry logic works.
+    def test_tsoAssignFailure(self):
+        scount, fcount = readMetrics()
+        self.assertEqual(scount, 1)
+        self.assertEqual(fcount, 2)
+        
+        time.sleep(1)
+        new_tso_args = shlex.split(args.cmd)
+        p = subprocess.Popen(new_tso_args)
+        time.sleep(5)
+
+        scount, fcount = readMetrics()
+        self.assertEqual(scount, 3)
+        self.assertEqual(fcount, 0)
+
+        p.kill()
```

Comment:
should also do p.wait after kill to make sure it exits, perhaps even check to make sure exit code is 0

Reply:
I have added the p.wait for the subprocess. However it seems like the exit code from the python subprocess module does not match the POSIX exit codes. I can only test if the process is terminated by the signal I sent. https://docs.python.org/2/library/subprocess.html#subprocess.Popen.returncode

Reply:
not sure i understand. i'd do:
`
rcode = p.wait()
self.assertEqual(rcode, 0)
`

Reply:
Yes this is exactly what i thought it should return but the `rcode` would be -9 instead of 0 since p.wait() just set and returns the `p.returncode` variable according to https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait. The variable is set to -9, which means it was stopped by a SIGKILL.

Reply:
oh i see. The problem is that you're using SIGKILL. You need to send SIGTERM for graceful shutdown

Reply:
I was using the posix speak for kill() which just means deliver a signal. It seems the Popen API has a separate send_signal() method to use instead of kill()

Reply:
Ahh that make sense. Thanks!